### ISSUE TITLE
[coor_refactor] fix TICA correlation tests

### ISSUE THREAD

**Issue by marscher** [markovmodel/PyEMMA#666]
TICA now uses the newly introduced covartools extension and the functional tests (test_feature_reader_and_tica*) are working as expected.

test_feature_correlation\* tests in test_tica are still now showing the expected results:

Some elements are off by orders of magnitude.

``` python
============================= ERRORS =============================
Traceback (most recent call last):
  File "/home/marscher/anaconda/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/marscher/workspace/pyemma/pyemma/coordinates/tests/test_tica.py", line 291, in test_feature_correlation_MD
    np.testing.assert_allclose(test_corr, true_corr, atol=1e-3)
  File "/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line 708, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.001

(mismatch 88.3198106829%)
 x: array([[ 0.025058,  0.029361, -0.083287, ..., -0.221143,  0.002966,
        -0.094069],
       [-0.069385, -0.018301, -0.155376, ..., -0.248272,  0.087544,...
 y: array([[ 0.01956 ,  0.031687, -0.074823, ..., -0.215951, -0.004571,
        -0.090331],
       [-0.066171, -0.028724, -0.157047, ..., -0.255688,  0.090109,...
-------------------- >> begin captured logging << --------------------
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: reset (previous) parametrization state, since data producer has been changed.
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: Running TICA with tau=1; Estimating two covariance matrices with dimension (174, 174)
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: using 2 moments for 1 chunks
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: diagonalize Cov and Cov_tau.
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: finished diagonalisation.
pyemma.coordinates.data.feature_reader.FeatureReader[4]: DEBUG: get_output(): dimensions=slice(0, None, None)
pyemma.coordinates.data.feature_reader.FeatureReader[4]: DEBUG: get_output(): created output trajs with shapes: [(100, 174)]
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: get_output(): dimensions=slice(0, None, None)
pyemma.coordinates.transform.tica.TICA[3]: DEBUG: get_output(): created output trajs with shapes: [(100, 68)]
--------------------- >> end captured logging << ---------------------
```

and 

``` python
============================= ERRORS =============================
Traceback (most recent call last):
  File "/home/marscher/anaconda/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/marscher/workspace/pyemma/pyemma/coordinates/tests/test_tica.py", line 309, in test_feature_correlation_data
    np.testing.assert_allclose(test_corr, true_corr)
  File "/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line 708, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

(mismatch 100.0%)
 x: array([[ 0.999946,  0.006831,  0.007849],
       [ 0.331673, -0.764548,  0.552683],
       [-0.041568,  0.504241,  0.862562]])
 y: array([[ 0.999955,  0.028576,  0.036958],
       [ 0.360446, -0.749952,  0.543483],
       [-0.008842,  0.519828,  0.860426]])
-------------------- >> begin captured logging << --------------------
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: reset (previous) parametrization state, since data producer has been changed.
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: Running TICA with tau=10; Estimating two covariance matrices with dimension (3, 3)
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: using 2 moments for 1 chunks
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: diagonalize Cov and Cov_tau.
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: finished diagonalisation.
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: get_output(): dimensions=slice(0, None, None)
pyemma.coordinates.transform.tica.TICA[5]: DEBUG: get_output(): created output trajs with shapes: [(100, 3)]
--------------------- >> end captured logging << ---------------------
```

**Comment by franknoe** [comment]
pls discuss with @fabian-paul or @fnueske

Am 06/01/16 um 19:13 schrieb Martin K. Scherer:

> TICA now uses the newly introduced covartools extension and the 
> functional tests (test_feature_reader_and_tica*) are working as expected.
> 
> test_feature_correlation\* tests in test_tica are still now showing the 
> expected results:
> 
> Some elements are off by orders of magnitude.
> 
> =============================  ERRORS=============================
> Traceback (most recent call last):
>    File"/home/marscher/anaconda/lib/python2.7/unittest/case.py", line329,in  run
>      testMethod()
>    File"/home/marscher/workspace/pyemma/pyemma/coordinates/tests/test_tica.py", line291,in  test_feature_correlation_MD
>      np.testing.assert_allclose(test_corr, true_corr,atol=1e-3)
>    File"/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line1347,in  assert_allclose
>      verbose=verbose, header=header)
>    File"/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line708,in  assert_array_compare
>      raise  AssertionError(msg)
> AssertionError:
> Not equal to tolerance rtol=1e-07, atol=0.001
> 
> (mismatch88.3198106829%)
>   x: array([[0.025058,0.029361,-0.083287, ...,-0.221143,0.002966,
>          -0.094069],
>         [-0.069385,-0.018301,-0.155376, ...,-0.248272,0.087544,...
>   y: array([[0.01956  ,0.031687,-0.074823, ...,-0.215951,-0.004571,
>          -0.090331],
>         [-0.066171,-0.028724,-0.157047, ...,-0.255688,0.090109,...
> --------------------  >>  begin captured logging<<  --------------------
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: reset (previous) parametrization state, since data producer has been changed.
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: Running TICAwith  tau=1; Estimating two covariance matriceswith  dimension (174,174)
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: using2  momentsfor  1  chunks
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: diagonalize Covand  Cov_tau.
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: finished diagonalisation.
> pyemma.coordinates.data.feature_reader.FeatureReader[4]: DEBUG: get_output(): dimensions=slice(0,None,None)
> pyemma.coordinates.data.feature_reader.FeatureReader[4]: DEBUG: get_output(): created output trajswith  shapes: [(100,174)]
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: get_output(): dimensions=slice(0,None,None)
> pyemma.coordinates.transform.tica.TICA[3]: DEBUG: get_output(): created output trajswith  shapes: [(100,68)]
> ---------------------  >>  end captured logging<<  ---------------------
> 
> and
> 
> =============================  ERRORS=============================
> Traceback (most recent call last):
>    File"/home/marscher/anaconda/lib/python2.7/unittest/case.py", line329,in  run
>      testMethod()
>    File"/home/marscher/workspace/pyemma/pyemma/coordinates/tests/test_tica.py", line309,in  test_feature_correlation_data
>      np.testing.assert_allclose(test_corr, true_corr)
>    File"/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line1347,in  assert_allclose
>      verbose=verbose, header=header)
>    File"/home/marscher/anaconda/lib/python2.7/site-packages/numpy/testing/utils.py", line708,in  assert_array_compare
>      raise  AssertionError(msg)
> AssertionError:
> Not equal to tolerance rtol=1e-07, atol=0
> 
> (mismatch100.0%)
>   x: array([[0.999946,0.006831,0.007849],
>         [0.331673,-0.764548,0.552683],
>         [-0.041568,0.504241,0.862562]])
>   y: array([[0.999955,0.028576,0.036958],
>         [0.360446,-0.749952,0.543483],
>         [-0.008842,0.519828,0.860426]])
> --------------------  >>  begin captured logging<<  --------------------
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: reset (previous) parametrization state, since data producer has been changed.
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: Running TICAwith  tau=10; Estimating two covariance matriceswith  dimension (3,3)
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: using2  momentsfor  1  chunks
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: diagonalize Covand  Cov_tau.
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: finished diagonalisation.
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: get_output(): dimensions=slice(0,None,None)
> pyemma.coordinates.transform.tica.TICA[5]: DEBUG: get_output(): created output trajswith  shapes: [(100,3)]
> ---------------------  >>  end captured logging<<  ---------------------
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by marscher** [comment]
We already tried to pin this down with @fabian-paul and @clonker, but we could not explain why the functional test succeeds and the correlation is wrong and the same time. We will take another look.

**Comment by franknoe** [comment]
ok, I can have a look next week. this week is pretty full.

Am 06/01/16 um 19:37 schrieb Martin K. Scherer:

> We already tried to pin this down with @fabian-paul 
> https://github.com/fabian-paul and @clonker 
> https://github.com/clonker, but we could not explain why the 
> functional test succeeds and the correlation is wrong and the same 
> time. We will take another look.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169413562.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by gph82** [comment]
Hi! May I suggest to compute the correlations directly with np.corr and see which values are the right ones? That should be doable if the input-feature trajectory and the TICA-trajectory are known, right?

**Comment by franknoe** [comment]
Also, since TICA is deterministic, should get identical results with old 
algorithm (devel-branch) and new algorithm (new branch) on fixed test 
data. I think the unit tests only test roughly because data is generated 
by a stochastic model there.

Am 06/01/16 um 19:37 schrieb Martin K. Scherer:

> We already tried to pin this down with @fabian-paul 
> https://github.com/fabian-paul and @clonker 
> https://github.com/clonker, but we could not explain why the 
> functional test succeeds and the correlation is wrong and the same 
> time. We will take another look.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169413562.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
Already the correlation matrix is quite off:

``` Python
from pyemma.coordinates import source, tica
import numpy as np

pdb_file = 'data/bpti_ca.pdb'
xtc_file = 'data/bpti_mini.xtc'
inp = source(xtc_file, top=pdb_file)
ticamini = tica(inp, lag=1, kinetic_map=False)

feature_traj =  ticamini.data_producer.get_output()[0]
tica_traj    =  ticamini.get_output()[0]

print np.cov(feature_traj.T)-ticamini.cov
```

```
[[-0.00477511 -0.01008754  0.00787608 ..., -0.00461037 -0.0103089
   0.01087911]
 [-0.01008754  0.00450424 -0.01375546 ..., -0.01148648  0.00806873
  -0.01766821]
 [ 0.00787608 -0.01375546 -0.00419868 ...,  0.01000304 -0.01587945
  -0.00079491]
 ..., 
 [-0.00461037 -0.01148648  0.01000304 ..., -0.00531534 -0.01194915
   0.01325944]
 [-0.0103089   0.00806873 -0.01587945 ..., -0.01194915  0.01206735
  -0.02018694]
 [ 0.01087911 -0.01766821 -0.00079491 ...,  0.01325944 -0.02018694
   0.00313061]]
```

**Comment by franknoe** [comment]
What about normalization? I think in covartools there is not necessarily 
a normalization, unless you request it, because for symmetric TICA you 
don't need to normalize the matrices.

Am 07/01/16 um 10:45 schrieb fabian-paul:

> Already the correlation matrix is quite off:
> 
> from  pyemma.coordinatesimport  source, tica
> import  numpyas  np
> 
> pdb_file=  'data/bpti_ca.pdb'
> xtc_file=  'data/bpti_mini.xtc'
> inp=  source(xtc_file,top=pdb_file)
> ticamini=  tica(inp,lag=1,kinetic_map=False)
> 
> feature_traj=   ticamini.data_producer.get_output()[0]
> tica_traj=   ticamini.get_output()[0]
> 
> print  np.cov(feature_traj.T)-ticamini.cov
> |[[-0.00477511 -0.01008754 0.00787608 ..., -0.00461037 -0.0103089 
> 0.01087911] [-0.01008754 0.00450424 -0.01375546 ..., -0.01148648 
> 0.00806873 -0.01766821] [ 0.00787608 -0.01375546 -0.00419868 ..., 
> 0.01000304 -0.01587945 -0.00079491] ..., [-0.00461037 -0.01148648 
> 0.01000304 ..., -0.00531534 -0.01194915 0.01325944] [-0.0103089 
> 0.00806873 -0.01587945 ..., -0.01194915 0.01206735 -0.02018694] [ 
> 0.01087911 -0.01766821 -0.00079491 ..., 0.01325944 -0.02018694 
> 0.00313061]] |
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169611858.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
Then how do I request normalization?

**Comment by franknoe** [comment]
I just checked, the cov_XX and cov_XY functions you are using do use 
normalization. Maybe you are normalizing differently in the reference?
symmetrize=True means we are using Fabian's symmetric TICA version, 
which I think corresponded to the previous force_ev_smaller_one = True.

If you think it might be a fundamental problem, please test directly in
     https://github.com/markovmodel/variational
which has a lot of test cases for the covar functions, but it might 
still be that I missed something.

Out of curiosity: l. 258 of
https://github.com/markovmodel/PyEMMA/blob/add_covar_ext/pyemma/coordinates/transform/tica.py
contains
`self.cov = self._cover.covXX()`

How does that even work? There is no function covXX(), just cov_XX() in 
the RunningCovar class.

Am 07/01/16 um 11:00 schrieb fabian-paul:

> Then how do I request normalization?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169614665.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
Actually the pepeline refactoring branch uses its own copy of the variational functions, see https://github.com/markovmodel/PyEMMA/tree/devel_refactor_transformer/pyemma/coordinates/estimators/covar
It this still intended or should we change this?

**Comment by franknoe** [comment]
yes, intended. There is no stable API for a subpackage yet.

Am 07/01/16 um 11:40 schrieb fabian-paul:

> Actually the pepeline refactoring branch uses its own copy of the 
> variational functions, see 
> https://github.com/markovmodel/PyEMMA/tree/devel_refactor_transformer/pyemma/coordinates/estimators/covar
> It this still intended or should we change this?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169625549.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
What is the development workflow? I see that there are new commits in that per-branch-copy that are not present in markovmodel/variational. Is this how it should work? Making changes in the per-branch-copy and later committing them to markovmodel/variational?

**Comment by franknoe** [comment]
Am 07/01/16 um 11:46 schrieb fabian-paul:

> What is the development workflow? I see that there are new commits in 
> that per-branch-copy that are not present in markovmodel/variational.
> 
> Why?
> 
> Is this how it should work?
> 
> No, if there is a problem in that code it should be committed to the 
> source, which is at markovmodel/variational.
> 
> Making changes in the per-branch-copy and later committing them to 
> markovmodel/variational?
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/markovmodel/PyEMMA/issues/666#issuecomment-169626675.

## 

---

Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de

## Mail:  Arnimallee 6, 14195 Berlin, Germany

**Comment by fabian-paul** [comment]
Fixed in commits bd38b78d3664c26c8723cd1bb42b24937722d06e, c932de0834b9a366806d9edf9efcd76e2306c104 and 258dcc421070168b1cb568b5878264d36f340cfa
